File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode node, int target) {
        if (node == null) {
            return false;
        }
        target -= node.val;
        if (node.left == null && node.right == null) {
            return target == 0;
        }

        return hasPathSum(node.left, target) || hasPathSum(node.right, target);
    }
}

File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], target: int) -> bool:
        if not root:
            return False
        stack = [(root, target)]
        while stack:
            node, n = stack.pop()
            if not node.left and not node.right and node.val - n == 0:
                return True
            if node.right:
                stack.append((node.right, n - node.val))
            if node.left:
                stack.append((node.left, n - node.val))
        return False